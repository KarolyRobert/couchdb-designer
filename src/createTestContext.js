import createSectionFromDirectory from './section/createSectionFromDirectory';
import { registerContext,addValidator,getTestContext } from '../build/testing/testEnvironment';
import {registerDatabase} from './testing/changes/updateDocument';
import createCouchDBFunctions from './testing/createCouchDBFunctions';
import defaults from './testing/defaults';
import contextFunction from './util/contextFunction';
//import testBuiltIns from './testing/testBuiltIns';
import crypto from 'crypto';
import path from 'path';

export default function createTestContext(directoryName,testDatabase,userCtx = defaults.userCtx,secObj = defaults.secObj,parentContext = false){
    if(process.env.JEST_WORKER_ID === undefined){
        throw new Error('createTestContext can only be used inside Jest Framework!');
    }
    return new Promise((resolve,reject) => {
        let testContext;
        let contextProps;
        let root = path.join(directoryName);
        let name = root.split(path.sep).pop();
        let directory = path.join(root, '..');
        if(parentContext){
            contextProps = {root,name,contextId:parentContext};  
            testContext = {};
        }else{
            let fullPath = path.resolve(process.env.PWD,root);
            let contextId = crypto.createHash('md5').update(fullPath).digest('hex');
         
            contextProps = {root,contextId}
            testContext = contextFunction(contextId); 
        }

        testContext.id = `_design/${name}`;
        testContext.language = 'javascript';
        

        createSectionFromDirectory(directory, name, contextProps).then(section => {
            
            testContext = Object.assign(testContext, section[name]);
         
            if(testContext.language.toLowerCase() === 'javascript'){
                if(testContext.validate_doc_update){
                    addValidator(contextProps.contextId,testContext.id,testContext.validate_doc_update);
                }
               
                if(!parentContext){
                    createCouchDBFunctions(contextProps.contextId, testContext);
              
                    registerContext(contextProps.contextId, testContext,'context', userCtx, secObj);
                    registerDatabase(contextProps.contextId,testDatabase,userCtx);
                }else{
                    createCouchDBFunctions(contextProps.contextId, testContext,name);
                }
                resolve(testContext);
            }else if(!parentContext){
                reject(`Only "javascript" type design document testing is supported yet. This directory structure defining one "${testContext.language}" type design document!`);
            }else{
                resolve(testContext);
            }
        },err => {
            reject(err)
        });
        
    });
}