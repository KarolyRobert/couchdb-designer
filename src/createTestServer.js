import fs from 'fs/promises';
import createTestContext from './createTestContext';
import createMangoContext from './createMangoContext';
import { registerContext } from '../build/testing/testEnvironment';
import contextFunction from './util/contextFunction';
import defaults from './testing/defaults';
import {registerDatabase} from './testing/changes/updateDocument';
import createAllDocs from './testing/indexes/createAllDocs';
import createMangoFunctions from './testing/createMangoFunctions';
import crypto from 'crypto';
import path from 'path';

const createTestServer = (directoryName,testDatabase,userCtx = defaults.userCtx,secObj = defaults.secObj) => {
    return new Promise((resolve,reject) => {
        let root = path.join(directoryName);
        let fullPath = path.resolve(process.env.PWD,root);
        let contextId = crypto.createHash('md5').update(fullPath).digest('hex');
      
        fs.readdir(root).then(names => {
            Promise.all(names.map(name => {
                if(/.*\.json$/.test(name.toLowerCase())){
                    return createMangoContext(directoryName,name,testDatabase.partitioned ? true : false,contextId);
                }else{
                    return createTestContext(path.join(directoryName,name),testDatabase,null,null,contextId);
                }
            }))
            .then(designContexts => {

                let serverContext = contextFunction(contextId);
                
                for(let designContext of designContexts){
                    if(designContext){
                        let designName = designContext.id.split('/')[1];
                    
                        serverContext[designName] = designContext;
                        
                    }
                }
                registerContext(contextId, serverContext,'server', userCtx, secObj);
                registerDatabase(contextId,testDatabase,userCtx);
                createAllDocs(contextId);
                createMangoFunctions(contextId);
                resolve(serverContext);
            },err => reject(err));
               
                
        },err => reject(err));

    });
}

export default createTestServer;