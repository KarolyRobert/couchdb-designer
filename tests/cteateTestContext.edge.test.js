import createTestContext from "../src/createTestContext";
import createTestServer from "../src/createTestServer";

const testDatabase = {name:'proba1',partitioned:true,data:[
    {_id:'p:doc1',name:[[10,10],9,13],parent:'roger'},
    {_id:'p:doc2',name:[[10,10],9,10],parent:{name:"zargu"}},
    {_id:'s:doc1',name:[[10,10],9,13],parent:'roger'},
    {_id:'s:doc2',name:[[10,10],9,10],parent:'zargu'}
    
]}

const testDatabase2 = {name:'proba2',partitioned:true,data:[
    {_id:'p:doc1',name:'zargu',parent:'roger'},
    {_id:'p:doc2',name:'roger',parent:'zargu'}
]}

    describe("createTestContext",() => {
        beforeEach(() => {
            jest.clearAllMocks();
        });
        test("all case",() => {
            return createTestContext('./tests/design/appdesign',testDatabase).then(context => {
                let update = context.updates.updateFromDir({},{});
                expect(update[0]).toEqual({updateByUpdateFromDir:'libfunction call updated'});
                expect(context.lib.couchdb.libfunction.mock.calls.length).toBe(1);
                expect(() => context.views.byDate.reduce(['keys'],['values'],false)).toThrow("Calling 'require' from reduce function in is not allowed and useless from library! You can fix it in tests/design/appdesign/views/byDate/reduce.js");
                context.views.byName.map(testDatabase.data[1]);
                expect(context('emitted')).toEqual({offset:0,total_rows:1,rows:[{id:'p:doc2',key:[[10,10],9,10],value:1}]});
                expect(context('logged')).toBe('[info] Log :: log from updateFromDir\n[info] Log :: log from views/byName/map\n');
            })//.catch(err => expect(err).toBe('nincs'));
        });
        
        test("concurent",() => {
            return createTestContext('./tests/design/appdesign',testDatabase2).then(context => {
                let update = context.updates.updateFromDir({},{});
                expect(update[0]).toEqual({updateByUpdateFromDir:'libfunction call updated'});
                expect(context.lib.couchdb.libfunction.mock.calls.length).toBe(1);
                expect(() => context.views.byDate.reduce(['keys'],['values'],false)).toThrow("Calling 'require' from reduce function in is not allowed and useless from library! You can fix it in tests/design/appdesign/views/byDate/reduce.js");
                expect(() => context('server').view.byName({group_level:1})).toThrow();
                expect(context('_design').update.updateFromDir({uuid:'uid1'},'p:doc1').__supplemented).toBeUndefined();
                expect(context('database')[0].updateByUpdateFromDir).toBe("libfunction call updated");
                expect(context('_changes',{filter:'appdesign/all'}).results.length).toBe(4);
            })
        });

        test('querys', () => {
            return createTestContext('./tests/design/querys',testDatabase).catch(err => {
                expect(err).toBe('Only "javascript" type design document testing is supported yet. This directory structure defining one "query" type design document!');
            });
        });
        test('server', () => {
            return createTestServer('./tests/design',testDatabase).then(server => {
                expect(server('server').appdesign.update.updateFromDir({uuid:'uid1'},'p:doc2').info.update_seq).toBe(4);
                expect(server('database')[1].updateByUpdateFromDir).toBe("libfunction call updated");
                expect(server('database','p:doc2').updateByUpdateFromDir).toBe("libfunction call updated");
                expect(server('database','p:doc1').parent).toBe("roger");
                expect(server('_changes',{filter:'appdesign/nothing'}).last_seq).toBe("4-AbRomqy2vFXcd0Yf4RijLw==");
                expect(() => server('_changes',{filter:'appdesig/nothing'})).toThrow('It is not appdesig design document in testdatabase!');
                expect(server('server').allDocs('p')).toMatchSnapshot();
                expect(server('server').find({selector:{parent:{"$gt":"aoger"}},sort:['parent'],use_index:['mango','mongo'],fields:['parent.name']},'s')).toEqual({docs:[{}]});
            });
        });
    });
