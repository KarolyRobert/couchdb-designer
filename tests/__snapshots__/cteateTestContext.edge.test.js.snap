// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTestContext server 1`] = `
Object {
  "offset": 0,
  "rows": Array [
    Object {
      "id": "p:doc1",
      "key": "p:doc1",
      "value": Object {
        "rev": "1-8a769907c5062489ba0c72727c6e906e",
      },
    },
    Object {
      "id": "p:doc2",
      "key": "p:doc2",
      "value": Object {
        "rev": "2-b07835dae7ff0c087c23d2b4b72709a8",
      },
    },
  ],
  "total_rows": 2,
}
`;
